examples/cookies-to-headers/src/main.rs

```diff
@@ -54,86 +48,83 @@ mod tests {
-                        .headers()
-                        .get("tasty_cookie")
-                        .expect("tasty_cookie is present"),
-                    "strawberry"
-                );
-                Ok(SubgraphResponse::fake_builder()
-                    .data(expected_mock_response_data)
-                    .build())
-            });
-
-        let conf: Conf = serde_json::from_value(serde_json::json!({
-            "scripts": "src",
-            "main": "cookies_to_headers.rhai",
-        }))
-        .expect("json must be valid");
-
-        // Build an instance of our plugin to use in the test harness
-        let rhai = Rhai::new(PluginInit::new(conf, Default::default()))
+        mock_service.expect_clone().return_once(move || {
+            let mut mock_service = test::MockSubgraphService::new();
+            mock_service
+                .expect_call()
+                .once()
+                .returning(move |req: subgraph::Request| {
+                    // Let's make sure our request contains our new headers
+                    assert_eq!(
+                        req.subgraph_request
+                            .headers()
+                            .get("yummy_cookie")
+                            .expect("yummy_cookie is present"),
+                        "choco"
+                    );
+                    assert_eq!(
+                        req.subgraph_request
+                            .headers()
+                            .get("tasty_cookie")
+                            .expect("tasty_cookie is present"),
+                        "strawberry"
+                    );
+                    req.context
+                        .insert("mock_data", expected_mock_response_data.to_owned())
+                        .unwrap();
+                    Ok(subgraph::Response::fake_builder().build())
+                });
+            mock_service
+        });
+
+        let config = serde_json::json!({
+            "rhai": {
+                "scripts": "src",
+                "main": "cookies_to_headers.rhai",
+            }
+        });
+        let test_harness = apollo_router::TestHarness::builder()
+            .configuration_json(config)
+            .unwrap()
+            .extra_subgraph_plugin(move |_, _| mock_service.clone().boxed())
+            .extra_router_plugin(|service| {
+                service
+                    .map_response(|mut response| {
+                        let mock_data = response.context.get("mock_data").unwrap();
+                        let body = response.response.body_mut();
+                        let dummy = futures::stream::empty().boxed();
+                        let stream = std::mem::replace(body, dummy);
+                        *body = stream
+                            .map(move |mut resp| {
+                                resp.data = mock_data.clone();
+                                resp
+                            })
+                            .boxed();
```

Comment:
is this why we need futures::StreamExt?

This feels a bit inconvenient, I m not sure I would assume plugin writers to be able to wield this, especially when learning rust :/

Reply:
Yes, this was an ugly way to get this test to compile. I’ve added a helper in https://github.com/apollographql/router/pull/1525